---
title: "MSB7102 Mini-project, Semester I, 2021"
author: "Arthur Shem Kasambula"
date: "6/25/2021"
output:
  html_document: default
  pdf_document: default
  word_document: default
---
# Background

The data used in this exercise is derived and was generated by Kolistic et al in their study “The dynamics of the human infant gut microbiome in development and progression toward Type 1 Diabetes”. See full publication at [here](https://doi.org/10.1016/j.chom.2015.01.001). Briefly, this was a prospective analysis of developing gut microbiome in infants en route to type 1 diabetes. Infants from Finland and Estonia were recruited at birth based on predisposition to autoimmunity determined by human leukocyte antigen (HLA) genotyping. The cohort consists of 33 infants, 11 of which seroconverted to serum autoantibody positivity and of those, four developed T1D within the three-year time-frame of this study.

# Tasks 

*1*. _Import the data described above into R, provide descriptive summaries of the subject data (using appropriate graphics and statistical summary measures) given in the diabimmune_16s_t1d_metadata.csv file_. _In addition, use appropriate test(s) to check for association/independency between disease status and other variables (delivery mode, gender and age)_. _Note that age is given in days_.

*2*. _Using phyloseq, create a phyloseq object. This will comprise the OTU abundance, taxonomy (provided in the .txt file) and sample data (provided in the .csv file)_.

*3*. _Generate Alpha diversity plots and ordination plots. Examine any observed patterns by delivery mode, gender and disease status_.

*4*. _Perform differential abundance using DEseq2_.

# Answers

```{r packages, eval=FALSE}
#loading the required packages
library(readr)
library(dplyr)
library(viridis)
library(tidyverse)
library(gtsummary)
library(phyloseq)
library(ggplot2)
library(DESeq2)
```


```{r required packages, include=FALSE, message=FALSE}
library(readr)
library(dplyr)
library(tidyverse)
library(gtsummary)
library(phyloseq)
library(ggplot2)
library(DESeq2)
library(ggpubr)
library(viridis)
theme_set(theme_bw())
scale_fill_discrete <- function(palname = "Paired", ...) {
  scale_fill_brewer(palette = palname, ...)
}
```

The files to be imported include the diabimmune_16s_t1d_metadata.csv and the diabimmune_t1d_16s_otu_table.txt
```{r data importation, include=FALSE, message=FALSE}
#Importing the data to be analysed
metadata <- read_csv("diabimmune_16s_t1d_metadata.csv")
#changing the data to a tibble format
metadata1 <- tbl_df(metadata)
taxa <- read_tsv("diabimmune_t1d_16s_otu_table.txt", skip = 1)
```


```{r cleaning the data, include=FALSE, message=FALSE}

#cleaning the data to create new dataframe grouped by subject_ID
by_subjectid <- group_by(metadata1, Subject_ID)
metadata2 <- summarize(by_subjectid,
                      Samples = n(),
                      Age = mean(Age_at_Collection))

metadata3 <- metadata1 %>% select(Subject_ID, Case_Control, Gender, Delivery_Route) %>% unique
#Merge the metadata2 and metadata3
metadata_cleaned <- metadata2 %>% inner_join(metadata3, by = "Subject_ID")
```

### Descriptive analysis
The total number of participants in the study were 33. Out of which 18(55%) were females and 15(45%) were males as shown in table1 below. Majority of the participants 30(91%) were delivered by the vaginal route and very few participants 3(9.1%) were delivered by cesarian section. Among the 33 only 11(33%) were cases. The median number of samples collected from each participant was 23 while the minimum and maximum number of samples collected from each participant was 8 and 34 respectively.

_**Table1**_ _**Showing characteristics of the participants**_
```{r descriptive analysis, echo=FALSE, message=FALSE}
metadata_cleaned %>% 
  select(Gender, Delivery_Route, Case_Control, Samples) %>% 
  tbl_summary(     
    statistic = list(all_continuous() ~ "{median} ({min},{max})",        
                     all_categorical() ~ "{n} ({p}%)"),   
    type   = all_categorical() ~ "categorical",                 
    label  = list(                                              
      Gender ~ "Gender",
      Delivery_Route ~ "Delivery route",
      Case_Control ~ "Disease status"))
```

## Samples collected
In this study over 777 samples were collected as seen in figure1 and Table2 below. 

```{r graphics code, include=FALSE, message=FALSE}
plot_data1 <- metadata1 %>%
  group_by(Case_Control) %>%
  summarise(totals = n(), samples_collected = round((totals / nrow(metadata1)*100)), 1)

sample_graph1 <- ggplot(plot_data1, aes(x = Case_Control, y = samples_collected, fill = Case_Control)) + geom_bar(stat = "identity", width = 0.7) + geom_text(label = paste0(plot_data1$samples_collected, "%"), vjust = -0.2, position = position_dodge(width = 0.9)) + geom_col(colour = "black") + theme_classic() + theme(legend.position = "none") + xlab("Disease status")

#Distribution of Delivery route
plot_data2 <- metadata1 %>%
  group_by(Delivery_Route) %>%
  summarise(totals = n(), samples_collected = round((totals / nrow(metadata1)*100)), 1)

sample_graph2 <- ggplot(plot_data2, aes(x = Delivery_Route, y = samples_collected, fill = Delivery_Route)) + geom_bar(stat = "identity", width = 0.7) + geom_text(label = paste0(plot_data2$samples_collected, "%"), vjust = -0.2, position = position_dodge(width = 0.9)) + geom_col(colour = "black") + theme_classic() + theme(legend.position = "none") + xlab("Delivery Route")

#Distribution by gender
plot_data3 <- metadata1 %>%
  group_by(Gender) %>%
  summarise(totals = n(), samples_collected = round((totals / nrow(metadata1)*100)), 1)

sample_graph3 <- ggplot(plot_data3, aes(x = Gender, y = samples_collected, fill = Gender)) + geom_bar(stat = "identity", width = 0.7) + geom_text(label = paste0(plot_data3$samples_collected, "%"), vjust = -0.2, position = position_dodge(width = 0.9)) + geom_col(colour = "black") + theme_classic() + theme(legend.position = "none") + xlab("Gender")

```


```{r graphs, echo=FALSE, message=FALSE, fig.align='center', out.width="70%"}
ggarrange(sample_graph1, sample_graph2, sample_graph3, 
          labels = c("A", "B", "C"),
          ncol = 3, nrow = 1)
```

**A.** _Samples collected by disease status_
**B.** _Samples collected by delivery route_
**C.** _Samples collected by gender_

**Table2** _Showing a summary of samples collected_

```{r summary_table, include=FALSE, message=FALSE}
summary_table <- metadata1 %>% select(Age_at_Collection, Gender, Delivery_Route, Case_Control) %>% 
tbl_summary(statistic = list(all_continuous() ~ "{median} ({min},{max})",        
            all_categorical() ~ "{n} ({p}%)"),   
            type   = all_categorical() ~ "categorical",                 
            label  = list(                                              
            Age_at_Collection ~ "Age (days)",                           
            Gender ~ "Gender",
            Delivery_Route ~ "Delivery route",
            Case_Control ~ "Disease status"))
```


```{r printig summary table, echo=FALSE, message=FALSE,fig.align='center'}
summary_table
```


**Distribution of age at sample collection**
```{r graphs for age distribution at collection, include=FALSE, message=FALSE}
A <- metadata1 %>% ggplot(aes(x = Age_at_Collection, y = Subject_ID, color = Case_Control))+geom_point()
B <- metadata1 %>% ggplot(aes(x=Case_Control, y=Age_at_Collection, fill=Gender))+
geom_boxplot() + scale_fill_viridis(discrete = TRUE, alpha=0.6) +
geom_jitter(color="black", size = 1, alpha=0.9) +
theme_classic() +theme(legend.position="right",plot.title = element_text(size=11))+xlab("Disease status")

```

_**Figure D**_
```{r graph for Age 1, echo=FALSE, message=FALSE, fig.align='center', out.width="70%"}
A
```


_**Figure E**_
```{r graph for Age 2, echo=FALSE, message=FALSE, fig.align='center', out.width="70%"}
B
```


**Figure** **D** _and_ **E** _showing distribution of age at sample collection_

```{r chi-square test, include=FALSE, message=FALSE}
chi_results <- metadata1 %>% 
select(Gender, Delivery_Route, Case_Control) %>%    
tbl_summary(by = Case_Control, label  = list(                                              
Gender ~ "Gender",
Delivery_Route ~ "Delivery route")) %>%  
add_p()
```

**Testing for independence between the variables**

The independence between disease status and other variables was tested using the chi-square test. This reveals that disease status is influenced by the delivery route (p<0.001) as shown in table 3 below.

**Table3** _Showing p-values_
```{r testing for independency, echo=FALSE, message=FALSE, fig.align='center'}
chi_results

```


```{r Alpha diversity, include=FALSE, message=FALSE}
otu <- taxa %>% select(1:778)
names(otu)[1] <- "OTU_ID"
otu <- column_to_rownames(otu, var = "OTU_ID")

#Extract the taxonomy data table
taxonomy <- taxa %>% select(`#OTU ID`, ConsensusLineage) %>%
separate(ConsensusLineage, c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species"), sep = ";")
names(taxonomy)[1] <- "OTU_ID"
taxonomy <- column_to_rownames(taxonomy, var = "OTU_ID")

#Check for similarity in row names
identical(rownames(taxonomy),rownames(otu))

#Change OTU and taxonomy data frames to matrices
otu.matrix <- as.matrix(otu)
taxa.matrix <- as.matrix(taxonomy)

# create phyloseq object
phy_object <- phyloseq(otu_table(otu.matrix,taxa_are_rows=TRUE),tax_table(taxa.matrix))

#Import metadata
metadata4 <- column_to_rownames(metadata, var = "Sample_ID")

#Merge ASV table and taxonomic table as a phyloseq object
head(sample_names(phy_object))
length(sample_names(phy_object))
length(rownames(metadata4)) #check if same number of samples in phy_object and metadata file
length(intersect(rownames(metadata4),sample_names(phy_object)))#check that the sample names match in all cases

# Assign the metadata to the phyloseq object 'phy_object' 
sample_data(phy_object) <- metadata4 
identical(sample_names(phy_object),colnames(otu_table(phy_object)))

#test whether objects are the same
str(sample_data(phy_object))

sample_data(phy_object)[,"Gender"] <- as.factor(unlist(sample_data(phy_object)[,"Gender"]))
sample_data(phy_object)[,"Case_Control"] <- as.factor(unlist(sample_data(phy_object)[,"Case_Control"]))
sample_data(phy_object)[,"Delivery_Route"] <- as.factor(unlist(sample_data(phy_object)[,"Delivery_Route"]))

reads <- sample_sums(phy_object) #number of reads per sample
reads
#Set (arbitrary) cutoff for number of acceptable reads/sample. 
length(which(reads<5000)) #all samples have at least 5000 reads
#Let's standardize sample read count so we can compare between samples
total = median(sample_sums(phy_object))#find median sample read count
standf = function(x, t=total) round(t * (x / sum(x)))#function to standardize to median sample read count
M.std = transform_sample_counts(phy_object, standf)#apply to phyloseq object

```

**Plots showing Alpha diversity**
```{r graphs for Alpha diversity and ordination, include=FALSE, message=FALSE, cache=TRUE}
d1 <- plot_richness(M.std,x = "Case_Control",color = "Gender",measures=c("Chao1", "Shannon"), 
                   title = paste0("Alpha diversity by Case control and gender, N=",nsamples(M.std)))

d2 <- plot_richness(M.std,x = "Case_Control",color = "Delivery_Route",measures=c("Chao1", "Shannon"), 
                   title = paste0("Alpha diversity by Case control and  delivery route, N=",nsamples(M.std)))

ordinates <- ordinate(M.std, "PCoA", "bray")
plot_ordinate1 <- plot_ordination(M.std, ordinates, color = "Case_Control", title = "Ordination plot by Disease Status")

plot_ordinate2 <- ordinates %>%
plot_ordination(M.std, .,color = "Gender", title = "Ordination plot by Gender") 

plot_ordinate3 <- ordinates %>%
plot_ordination(M.std, ., color = "Delivery_Route", title = "Ordination plot by Delivery_Route")

```


```{r graph1 showing diversity, echo=FALSE, message=FALSE, fig.align='center', out.width="70%"}
d1
```

```{r graph2 showing diversity, echo=FALSE, message=FALSE, fig.align='center', out.width="70%"}
d2
```



**Ordination plots**

```{r plot1 for ordination, echo=FALSE, message=FALSE, fig.align='center', out.width="70%"}
plot_ordinate1
```


```{r plot2 for ordination, echo=FALSE, message=FALSE, fig.align='center', out.width="70%"}
plot_ordinate2
```


```{r plot3 for ordination, echo=FALSE, message=FALSE, fig.align='center', out.width="70%"}
plot_ordinate3
```


**Differential abundance**
```{r Differential abundance using deseq, include=FALSE, message=FALSE}
#DESeq: Creating a DESeq object
phy_object1 <- phyloseq(otu_table(otu.matrix + 1,taxa_are_rows=TRUE),tax_table(taxa.matrix))
sample_data(phy_object1) <- metadata4
sample_data(phy_object1)[,"Gender"] <- as.factor(unlist(sample_data(phy_object1)[,"Gender"]))
#sample_data(phy_object1)[,"Case_Control"] <- relevel(unlist(sample_data(phy_object1)[,"Case_Control"]), "control")
sample_data(phy_object1)[,"Delivery_Route"] <- as.factor(unlist(sample_data(phy_object1)[,"Delivery_Route"]))
casecontrol = phyloseq_to_deseq2(phy_object1, ~ Case_Control)

#DESeq test
casecontrol = DESeq(casecontrol, test="Wald", fitType="parametric")

#Results table
res = results(casecontrol, cooksCutoff = FALSE)
alpha = 0.01
casetab = res[which(res$padj < alpha), ]
casetab = cbind(as(casetab, "data.frame"), as(tax_table(phy_object1)[rownames(casetab), ], "matrix"))


# Phylum order
x = tapply(casetab$log2FoldChange, casetab$Phylum, function(x) max(x))
x = sort(x, TRUE)
casetab$Phylum = factor(as.character(casetab$Phylum), levels=names(x))
# Genus order
x = tapply(casetab$log2FoldChange, casetab$Genus, function(x) max(x))
x = sort(x, TRUE)
casetab$Genus = factor(as.character(casetab$Genus), levels=names(x))
diff_graph <- ggplot(casetab, aes(x=Genus, y=log2FoldChange, color=Phylum)) + geom_point(size=2) + 
theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5), legend.position = "right")

```


**Graph showing differential abundance**
```{r graph for differential abundance, echo=FALSE, message=FALSE, out.width="100%", fig.align='center'}
diff_graph
```





